cwl/src/main/scala/cwl/CwlWomExpression.scala

```diff
@@ -15,37 +20,66 @@ sealed trait CwlWomExpression extends WomExpression {
   override def evaluateValue(inputValues: Map[String, WomValue], ioFunctionSet: IoFunctionSet): ErrorOr[WomValue] = {
     val parameterContext = ParameterContext.Empty.withInputs(inputValues, ioFunctionSet)
 
-    val womValue = outputBinding.commandOutputBindingToWdlValue(parameterContext, ioFunctionSet)
-    cwlExpressionType.coerceRawValue(womValue).toErrorOr
+    val wdlValue: WomValue = outputBinding.commandOutputBindingToWdlValue(parameterContext, ioFunctionSet)
+    val extractFile: WomValue =
+      wdlValue match {
+        case WomArray(_, Seq(WomMap(WomMapType(WomStringType, WomStringType), map))) => map(WomString("location"))
+        case other => other
+      }
+    cwlExpressionType.coerceRawValue(extractFile).toErrorOr
   }
 
-  override def inputs: Set[String] = ???
-
   /*
   TODO:
    DB: It doesn't make sense to me that this function returns type WdlFile but accepts a type to which it coerces.
    Wouldn't coerceTo always == WdlFileType, and if not then what?
    */
-  override def evaluateFiles(inputTypes: Map[String, WomValue], ioFunctionSet: IoFunctionSet, coerceTo: WomType): ErrorOr[Set[WomFile]] ={
+  override def evaluateFiles(inputs: Map[String, WomValue], ioFunctionSet: IoFunctionSet, coerceTo: WomType): ErrorOr[Set[WomFile]] ={
+
+    val pc = ParameterContext.Empty.withInputs(inputs, ioFunctionSet)
 
-    val pc = ParameterContext.Empty.withInputs(inputTypes, ioFunctionSet)
-    val womValue = outputBinding.commandOutputBindingToWdlValue(pc, ioFunctionSet)
+    outputBinding.glob.toList.flatMap {
+      globValue =>
+        GlobEvaluator.globPaths(globValue, pc, ioFunctionSet).toList
+    }.map{
+      s:String =>
+        WomGlobFile(s): WomFile
+    }.toSet.validNel[String]
+  }
+}
 
-    womValue match {
+case class WorkflowStepInputExpression(input: WorkflowStepInput, override val cwlExpressionType: WomType, val graphInputs: Set[String]) extends CwlWomExpression {
 
-      case WomArray(WomMaybeEmptyArrayType(WomMapType(WomStringType, WomStringType)), seq: Seq[WomValue]) =>
-        seq.map {
-          case WomMap(WomMapType(WomStringType, WomStringType), map) => WomGlobFile(map(WomString("location")).valueString): WomFile
-        }.toSet.validNel
+  override def sourceString = input.toString
 
-      case other =>s":( we saw $other and couldn't convert to a globfile type: ${other.womType} coerceTo: $coerceTo".invalidNel[Set[WomFile]]
+  override def evaluateValue(inputValues: Map[String, WomValue], ioFunctionSet: IoFunctionSet) = {
+    (input.valueFrom, input.source) match {
+      case (None, Some(Inl(id: String))) =>
+        inputValues.
+          get(FullyQualifiedName(id).id).
+          toValidNel(s"could not find id $id in typeMap\n${inputValues.mkString("\n")}\nwhen evaluating $input.  Graph Inputs were ${graphInputs.mkString("\n")}")
+      case _ => Invalid(NonEmptyList.one("could not decipher evaluateValue, most likely has not been implemented yet"))
```

Comment:
`"could not decipher evaluateValue, most likely has not been implemented yet".invalidNel`